### ISSUE TITLE
Softmax scores

### ISSUE THREAD

**Issue by KoubaPetr** [kenziyuliu/MS-G3D#42]
Hi!

First, let me congratulate you on the work and a nice to use repo! I used your kinetics-pretrained model for finetuning and making a 10 class classifier of new activities and it works quite nice. I would just like to clarify one thing the output of your model (of the forward function) is the output of the last linear layer, right? So if I want to know the class probabilities, I still need to apply nn.Softmax() to the output, right? And in the ensemble, you again add the outputs of the linear layer before applying any loss/softmax, correct?

I am just confused that when I apply Softmax to the output of my fine-tuned model, the model tends to be overly confident (one prob nearing 1 and the rest nearing 0), so I just wanted to clarify this, so I could search for problems in my training, if my assumptions about your model are right.

Thank you in advance for any response!
Best,
Petr

**Comment by Anirudh257** [comment]
@KoubaPetr I think that you can use the MS-G3D model as it is and replace the final softmax layer with the new number of classes. You can try multiple things here by adding a new FC + softmax and try different fine-tuning heuristics.

**Comment by KoubaPetr** [comment]
@Anirudh257 Thanks for reply! Sorry I might have not posed my question right, I did the fine-tuning with my number of classes, there is no issue in that. What I am asking is about the original model (since my fine-tuned version only differs by number of neurons in the FC layer), whether I am reading the model right. When I call the model to infer on new data, lets say model(data), it returns the output of the Model.forward(data) function and the output of that is the output of the FC layer right? There is no softmax sofar, since it is hidden inside the loss function - right? The output of the FC layer is in terms of ordering equivalent to output of softmax, since softmax preserves the ordering, but if I want the probabilities I still need to apply Softmax manually, right?

**Comment by Anirudh257** [comment]
@KoubaPetr I get your point. You can change that FC to reflect the number of classes in your new dataset, I feel.

**Comment by kenziyuliu** [comment]
Thanks @KoubaPetr for your interest and @Anirudh257 for your inputs.

> So if I want to know the class probabilities, I still need to apply nn.Softmax() to the output, right?

Yes, the model by default outputs the *logits* (unnormalized predictions) without softmax. It can be used directly with `nn.CrossEntropyLoss`. Note that `argsort(logits) == argsort(softmax(logits))` so we don't need softmax for retrieving a prediction. For class probabilities (normalized prediction) you apply a softmax.

> And in the ensemble, you again add the outputs of the linear layer before applying any loss/softmax, correct?

I don't remember exactly but I think the ensembles add the logits rather than the probabilities. Though adding probabilities are also doable.

Hope this helps! Feel free to re-open or comment below for more questions.

**Comment by KoubaPetr** [comment]
@kenziyuliu Exactly what I needed, perfect! Thank you very much!
